### PR TITLE
Fix #10868: Handle semantic name in Java source code

### LINKED ISSUE #10868: Mixed compileOrder error with scala 3.0.0-M3 but success with 2.13.4
## Minimized code

```Scala
% mkdir -p project src/main/{scala,java}
% echo 'addSbtPlugin("ch.epfl.lamp" % "sbt-dotty" % "0.5.1")' > project/plugins.sbt
% echo 'crossScalaVersions := Seq("2.13.4", "3.0.0-M3")' > build.sbt
% echo 'trait T { object O }' > src/main/scala/T.scala
% echo 'class J { T.O$ o; }' > src/main/java/J.java
% sbt ";clean ;++2.13.4 compile"
// success
% sbt ";clean ;++3.0.0-M3 compile"
[error] -- [E006] Not Found Error: /tmp/src/main/java/J.java:1:10 ------------
[error] 1 |class J { T.O$ o; }
[error]   |          ^
[error]   |          Not found: T

% echo 'compileOrder := CompileOrder.ScalaThenJava' >> build.sbt
% sbt ";clean ;++3.0.0-M3 compile"
// success
```

## Output
Compile error (see above) in `Mixed` compileOrder with scala 3.0.0-M3 but success with 2.13.4

## Expectation
Compile successfully as in scala 2.

## Notes
+ `Minimized code` in this issue is same as in #9200 but I think this is a difference issue, maybe related to sbt-dotty?
+ Same errors in java 8 & 11
```scala
% java -version
openjdk version "1.8.0_272"
OpenJDK Runtime Environment (AdoptOpenJDK)(build 1.8.0_272-b10)
OpenJDK 64-Bit Server VM (AdoptOpenJDK)(build 25.272-b10, mixed mode)
% eval "$(cs java --env --jvm 11)"
% java -version
openjdk version "11.0.9" 2020-10-20
OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.9+11)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 11.0.9+11, mixed mode)
```

### PR DESCRIPTION
**Pull request by liufengyun**
Fix #10868: Handle semantic name in Java source code

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/typer/Typer.scala [hunk 1]
```diff
@@ -596,13 +596,30 @@ class Typer extends Namer
...
 
-    def typeSelectOnType(qual: untpd.Tree)(using Context) =
-      typedSelect(untpd.cpy.Select(tree)(qual, tree.name.toTypeName), pt)
+    def javaSelectOnType(qual: Tree)(using Context) =
+      // semantic name conversion for `O$` in java code
+      if !qual.symbol.is(JavaDefined) && tree.name.endsWith("$")
+      then
+        val nameOrig = tree.name.asSimpleName
+        val nameSemantic = nameOrig.slice(0, nameOrig.length - 1).moduleClassName
+        val tree2 = untpd.cpy.Select(tree)(qual, nameSemantic)
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/typer/Typer.scala:605]: unmangleClassName in NameOps already has similar logic

**Review by smarter — CHANGES_REQUESTED** [review]

File: compiler/src/dotty/tools/dotc/typer/Typer.scala (hunk 1 above)
Comment (liufengyun) [reply compiler/src/dotty/tools/dotc/typer/Typer.scala]: Thanks for the tip 👍

File: compiler/src/dotty/tools/dotc/typer/Typer.scala [hunk 2]
```diff
@@ -596,13 +596,28 @@ class Typer extends Namer
         .withSpan(tree.span)
         .computeNullable()
 
-    def typeSelectOnType(qual: untpd.Tree)(using Context) =
-      typedSelect(untpd.cpy.Select(tree)(qual, tree.name.toTypeName), pt)
+    def javaSelectOnType(qual: Tree)(using Context) =
+      // semantic name conversion for `O$` in java code
+      if !qual.symbol.is(JavaDefined)
+      then
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/typer/Typer.scala:602]: we usually put the `then` on the same line as the `if`

**Review by smarter — APPROVED** [review]

File: compiler/src/dotty/tools/dotc/typer/Typer.scala (hunk 2 above)
Comment (liufengyun) [reply compiler/src/dotty/tools/dotc/typer/Typer.scala]: Indeed, that looks better

**Review by smarter — APPROVED** [review]

**Pull request merged**